zerocopy-derive/src/lib.rs

```diff
@@ -634,6 +635,78 @@ fn derive_from_zeros_struct(ast: &DeriveInput, strct: &DataStruct) -> proc_macro
 // - all of its variants are fieldless
 // - one of the variants has a discriminant of `0`
 
+// Returns `Ok(index)` if variant `index` of the enum has a discriminant of
+// zero. If `Err(bool)` is returned, the boolean is true if the enum has unknown
+// discriminants (e.g. discriminants set to const expressions which we can't
+// evaluate in a proc macro). If the enum has unknown discriminants, then it
+// might have a zero variant that we just can't detect.
+fn find_zero_variant(enm: &DataEnum) -> Result<usize, bool> {
+    // Discriminants can be anywhere in the range [isize::MIN, usize::MAX]
+    // because the discriminant type may be signed or unsigned. Since we only
+    // care about tracking the discriminant when it's less than or equal to
+    // zero, we can avoid usize -> isize conversions and bounds checking by
+    // making the "next discriminant" value implicitly negative.
+    let mut next_negative_discriminant = Some(0usize);
+
+    // Sometimes we encounter explicit discriminants that we can't know the
+    // value of (e.g. a constant expression that requires evaluation). These
+    // could evaluate to zero or a negative number, but we can't assume that
+    // they do (no false positives allowed!). So we treat them like strictly-
+    // positive values that can't result in any zero variants, and track whether
+    // we've encountered any unknown discriminants.
+    let mut has_unknown_discriminants = false;
+
+    for (i, v) in enm.variants.iter().enumerate() {
+        match v.discriminant.as_ref() {
+            // Implicit discriminant
+            None => {
+                match next_negative_discriminant.as_mut() {
+                    Some(0) => return Ok(i),
+                    // n is nonzero so subtraction is always safe
+                    Some(n) => *n -= 1,
+                    None => (),
+                }
+            }
+            // Explicit positive discriminant
+            Some((_, Expr::Lit(ExprLit { lit: Lit::Int(int), .. }))) => {
+                match int.base10_parse::<usize>().ok() {
+                    Some(0) => return Ok(i),
+                    Some(_) => next_negative_discriminant = None,
+                    None => {
+                        has_unknown_discriminants = true;
+                        next_negative_discriminant = None;
+                    }
```

Comment:
Under what conditions can we arrive here? Wouldn't it just be a case where Rust has permitted, e.g., `#[repr(u128)]`, and the user has provided a discriminant which can't fit in `usize`? It must at least be a valid positive number since it's inside a `Lit::Int`.

If that's right, then we should consider perhaps expanding the error messages we can emit to mentioning this case, or maybe even tracking "a number we couldn't parse" as a separate condition.

Reply:
`#[repr(u128)]` enums are unstable: https://github.com/rust-lang/rust/issues/56071. I modified this to use `u64` instead of `usize` as a result. I think this should cover all the cases where we could fail to parse, so this branch shouldn't be reachable in practice. That said, I think we'd prefer to fail gracefully rather than error here.

Reply:
True, but if it is stabilized, then this code may be used with `#[repr(u128)]` enums. Maybe we could just use `u128`s here (and document that it's for forwards-compatibility with https://github.com/rust-lang/rust/issues/56071)? I'm not worried about `#[repr(u256)]` enums, especially considering that this is just an issue of how nice our error messages are, not a correctness issue.

Reply:
Sorry, waiting on this change or a response to this comment before approving.

Reply:
Sure, bumped it up to 128 bits